elastica/experimental/connection_contact_joint/generic_system_type_connection.py

```diff
@@ -0,0 +1,396 @@
+        relative_velocity = velocity_system_two - velocity_system_one
+
+        # Compute damping force considering the specified damping coefficient `nu`
+        damping_force = self.nu * relative_velocity
+
+        # compute contact force as addition of elastic force and damping force
+        contact_force = elastic_force + damping_force
+
+        # loop over the two systems
+        for i, (system, index, point, system_position) in enumerate(
+            zip(
+                [system_one, system_two],
+                [index_one, index_two],
+                [self.point_system_one, self.point_system_two],
+                [position_system_one, position_system_two],
+            )
+        ):
+            # The external force has opposite signs for the two systems:
+            # For system one: external_force = +contact_force
+            # For system two: external_force = -contact_force
+            external_force = (1 - 2 * i) * contact_force
+
+            # the contact force needs to be applied at a distance from the Center of Mass (CoM) of the rigid body
+            # or the selected node of the Cosserat rod.
+            # This generates a torque, which we also need to apply to both systems.
+            # We first compute the vector r from the node / CoM to the joint connection point.
+            distance_system_point = (
+                system_position - system.position_collection[..., index]
+            )
+            # The torque is the cross product of the distance vector and the contact force: tau = r x F
+            external_torque = np.cross(distance_system_point, external_force)
+
+            # Apply external forces and torques to both systems.
+            system.external_forces[..., index] += external_force
+            # the torque still needs to be rotated into the local coordinate system of the system
+            system.external_torques[..., index] += (
+                system.director_collection[..., index] @ external_torque
+            )
+
+        return
+
+    def apply_torques(self, system_one, index_one, system_two, index_two):
+        """
+        Apply restoring joint torques to the connected systems.
+
+        In FreeJoint class, this routine simply passes.
+
+        Parameters
+        ----------
+        system_one : SystemType
+             System two of the joint connection.
+             Object of a child class of  either `CosseratRod` or `RigidBodyBase`.
+        index_one : int
+            Index of first system for joint.
+        system_two : SystemType
+            System two of the joint connection.
+            Object of a child class of  either `CosseratRod` or `RigidBodyBase`.
+        index_two : int
+            Index of second system for joint.
+
+        Returns
+        -------
+
+        """
+        pass
+
+
+class GenericSystemTypeFixedJoint(GenericSystemTypeFreeJoint, FixedJoint):
```

Comment:
@skim0119 Thanks for your comment. I don't think your suggestion would work, as we need to access the `apply_torques` method of the `FixedJoint` class. We also need to inherit from `GenericSystemTypeFreeJoint` to access their `apply_forces` method. This `GenericSystemTypeFixedJoint` class is basically a combination of `GenericSystemTypeFreeJoint` and `FixedJoint`.

Please let me know if you have another idea how to solve this.

Reply:
> I understand your point, but I feel it's little dangerous using multiple inheritances here. Those modules are not designed to be used as a mixin, and it can cause issues downstream.
> 
> For now, if you just want to assign the same method, simply put
> 
> ```python
> apply_torques = FixedJoint.apply_toruqes
> apply_forces = GenericSystemTypeFreeJoint.apply_forces
> ```
> 
> inside the class, instead of inheriting. You can also call `__init__` without inheriting, so your `__init__` method doesn't have to be changed.
> 
> > As you can see, the inheritance doesn't mean anything in your current implementation: it has a chance to be misinterpreted in later usage. Directly calling `GenericSystemTypeFreeJoint` and `FixedJoint` doesn't mean you are calling the parent entities; they are just calling other class members as static.
> 
> It appears to me that this class will eventually be converted to `FixedJoint` in the future, so let's keep the code as intuitive as possible.

This doesn't work, as the `__init__` cannot be called without inheriting. For the code
```py
FixedJoint.__init__(
            self, k=k, nu=nu, kt=kt, nut=nut, rest_rotation_matrix=rest_rotation_matrix
        )
```
I get the following error:
```
Traceback (most recent call last):
  File "/Users/maximilianstoelzle/Documents/phd/PyElastica/elastica/wrappers/connections.py", line 278, in __call__
    return self._connect_cls(*self._args, **self._kwargs)
  File "/Users/maximilianstoelzle/Documents/phd/PyElastica/elastica/experimental/connection_contact_joint/generic_system_type_connection.py", line 318, in __init__
    FixedJoint.__init__(
  File "/Users/maximilianstoelzle/Documents/phd/PyElastica/elastica/joint.py", line 240, in __init__
    super().__init__(k, nu)
TypeError: super(type, obj): obj must be an instance or subtype of type

```

Reply:
@skim0119  In my mind we have four options:

1. You know of a solution how to call the (__init__) methods if non-inherited classes 
2. We leave the inheritance like it is now. I think the syntax is quite clear how we leverage the functionality of both base classes we inherit from. I don't think it will cause any issues as a) it is an experimental feature anyway, so interface and functionality changes need to be expected b) As you mentioned the code should hopefully be merged / replace the FixedJoint class. 
3. We just copy the necessary code from the FixedJoint class into the GenericSystemTypeFixedJoint class. 
4. We created a static method in the FixedJoint class, which can also be called from the GenericSystemTypeJoint class without inheriting.

@skim0119 Please let me know which way you want to go.

Reply:
@mstoelzle 
I guess its a bit tricky. For now, let's just copy the code for `__init__`. I'll clear the code when we merge to `joint.py`. In the end, I think the sequence needs to be modularized.

For `apply_force` and `apply_torques`, I think it shouldn't cause an issue, right?

## Suggestions:

- Use `FreeJoint` as a base
- Have its own `__init__` function
- Make alias for `apply_force` and `apply_torques`

Reply:
> @mstoelzle I guess its a bit tricky. For now, let's just copy the code for `__init__`. I'll clear the code when we merge to `joint.py`. In the end, I think the initial computation needs to be modularized.
> 
> For `apply_force` and `apply_torques`, I think it shouldn't cause an issue, right?
> 
> ## Suggestions:
> * Use `FreeJoint` as a base
> * Have its own `__init__` function
> * Make alias for `apply_force` and `apply_torques`

@skim0119 I removed the inheritance from `FixedJoint` in https://github.com/GazzolaLab/PyElastica/pull/149/commits/80d0ed1511d41f457b4f6eb5994f6d0a3f8b661e. The `GenericSystemTypeFixedJoint` now uses `GenericSystemTypeFreeJoint` as a base class so that we can use their `__init__` without copying the code. The necessary `__init__` functionality from `FixedJoint` is copied.